ipa-core/src/net/transport.rs

```diff
@@ -43,68 +64,84 @@ impl HttpTransport {
         server_config: ServerConfig,
         network_config: NetworkConfig,
         clients: [MpcHelperClient; 3],
-        callbacks: TransportCallbacks<Arc<HttpTransport>>,
+        handler: Option<HandlerRef>,
     ) -> (Arc<Self>, MpcHelperServer) {
-        let transport = Self::new_internal(identity, clients, callbacks);
+        let transport = Self::new_internal(identity, clients, handler);
         let server = MpcHelperServer::new(Arc::clone(&transport), server_config, network_config);
         (transport, server)
     }
 
     fn new_internal(
         identity: HelperIdentity,
         clients: [MpcHelperClient; 3],
-        callbacks: TransportCallbacks<Arc<HttpTransport>>,
+        handler: Option<HandlerRef>,
     ) -> Arc<Self> {
         Arc::new(Self {
             identity,
-            callbacks,
             clients,
+            handler,
             record_streams: StreamCollection::default(),
         })
     }
 
-    pub fn receive_query(self: Arc<Self>, req: QueryConfig) -> ReceiveQueryResult {
-        (Arc::clone(&self).callbacks.receive_query)(self, req)
-    }
-
-    pub fn prepare_query(self: Arc<Self>, req: PrepareQuery) -> PrepareQueryResult {
-        (Arc::clone(&self).callbacks.prepare_query)(self, req)
-    }
-
-    pub fn query_input(self: Arc<Self>, req: QueryInput) -> QueryInputResult {
-        (Arc::clone(&self).callbacks.query_input)(self, req)
-    }
-
-    pub fn query_status(self: Arc<Self>, query_id: QueryId) -> QueryStatusResult {
-        (Arc::clone(&self).callbacks.query_status)(self, query_id)
-    }
-
-    pub fn complete_query(self: Arc<Self>, query_id: QueryId) -> CompleteQueryResult {
+    /// Dispatches the given request to the [`RequestHandler`] connected to this transport.
+    ///
+    /// ## Errors
+    /// Returns an error, if handler rejects the request for any reason.
+    ///
+    /// ## Panics
+    /// This will panic if request handler hasn't been previously set for this transport.
+    pub async fn dispatch<Q: QueryIdBinding, R: RouteParams<RouteId, Q, NoStep>>(
+        self: Arc<Self>,
+        req: R,
+        body: BodyStream,
+    ) -> Result<HelperResponse, ApiError>
+    where
+        Option<QueryId>: From<Q>,
+    {
         /// Cleans up the `records_stream` collection after drop to ensure this transport
         /// can process the next query even in case of a panic.
-        struct ClearOnDrop {
+        ///
+        /// This implementation is a poor man's safety net and only works because we run
+        /// one query at a time and don't use query identifiers.
```

Comment:
Is there an issue to improve this? Besides the issues you note, it seems like it might belong in the app or query processor rather than in (a particular implementation of) the transport layer.

Reply:
I am not sure where I want to put this - it feels like transport internal state must be self-managed. We currently share the same transport instance across all queries and therefore having a need to clean up the internal state. As anything can panic at any moment, this state management is repetitive and prone to errors.

Maybe a better model would create a transport per query and then any panic will cause thread to abort, gateway and request handler to be dropped and transport destroyed.